rdrf/rdrf/forms/widgets/settings_widgets.py

```diff
@@ -1,242 +1,151 @@
-        rows = [
-            self.generate_input('min', _('Min value'), parsed, _("leave empty if you want to use the CDE's min value")),
-            self.generate_input('max', _('Max value'), parsed, _("leave empty if you want to use the CDE's max value")),
-            self.generate_input('left_label', _('Left label'), parsed),
-            self.generate_input('right_label', _('Right label'), parsed),
-            self.generate_input('step', _('Step'), parsed),
-        ]
-        return "<br/>".join(rows)
+    def get_allowed_fields(self):
+        pass
 
-    def render(self, name, value, attrs=None, renderer=None):
-        parsed = {}
+    def parse_value(self, value):
+        self.parsed = {}
         try:
-            parsed = json.loads(value)
+            self.parsed = json.loads(value)
+            self.parsed = {k: v for k, v in self.parsed.items() if k in self.get_allowed_fields()}
         except Exception:
             pass
 
-        html = """
-             <div style="display:inline-grid" id="id_{name}">
-                {inputs}
-                <input type="hidden" name="{name}" value='{value}'/>
-             </div>""".format(inputs=self.generate_inputs(parsed), name=name, value=value)
-        javascript = """
-            function saveJSON() {
-                var inputs = $('#id_%s input[type!=hidden]');
-                var obj = {};
-                for (var i = 0; i < inputs.length; i++) {
-                    if (inputs[i].value !='' && inputs[i].value.trim() != '') {
-                        obj[inputs[i].name] = inputs[i].value;
-                    }
-                }
-                $("input[name='%s']").val(JSON.stringify(obj));
-            }
-            saveJSON();
-        """ % (name, name)
-        return mark_safe(f"""
-            {html}
-            <script>
-                {javascript}
-            </script>""")
+    def _update_input(self, name, title, info=None, onchange=None):
+        value = self.parsed.get(name, '')
+        self.parsed[name] = {
+            'name': name,
+            'value': value,
+            'title': title,
+            'info': info,
+            'onchange': onchange or 'saveJSON()'
+        }
+
+    def generate_text_input(self, name, title, **kwargs):
+        self._update_input(name, title, kwargs.get('info'), kwargs.get('onchange'))
+        self.parsed[name]['input_type'] = 'text'
+
+    def generate_checkbox_input(self, name, title, **kwargs):
+        self._update_input(name, title, kwargs.get('info'), kwargs.get('onchange'))
+        self.parsed[name]['input_type'] = 'checkbox'
+        self.parsed[name]['checked'] = kwargs.get('checked')
+
+    def generate_select_input(self, name, title, **kwargs):
+        self._update_input(name, title, kwargs.get('info'), kwargs.get('onchange'))
+        self.parsed[name]['input_type'] = 'select'
+        self.parsed[name]['options'] = kwargs.get('options', [])
+
+    def generate_inputs(self):
+        pass
```

Comment:
Same comment as above. `NotImplementedError` and move it up next to `get_allowd_fields` so it is clear for subclasses what they will have to override.

Reply:
Updated.